packages/backend/src/services/ExcelService/ExcelService.ts

```diff
@@ -303,71 +234,237 @@ export class ExcelService {
-                reject(error);
-            });
+        // Commit Excel to temp file
+        worksheet.commit();
+        await workbook.commit();
+
+        return rowCount;
+    }
+
+    /**
+     * Direct Excel export using streaming to minimize memory usage
+     * Processes JSONL data row-by-row and streams directly to S3
+     */
+    static async downloadAsyncExcelDirectly(
+        resultsFileName: string,
+        fields: ItemsMap,
+        storageClient: S3ResultsFileStorageClient,
+        options: {
+            onlyRaw?: boolean;
+            showTableNames?: boolean;
+            customLabels?: Record<string, string>;
+            columnOrder?: string[];
+            hiddenFields?: string[];
+            attachmentDownloadName?: string;
+        } = {},
+    ): Promise<{ fileUrl: string; truncated: boolean }> {
+        // Generate a unique filename
+        const formattedFileName = ExcelService.generateFileId(resultsFileName);
+
+        // Handle column ordering and filtering
+        const {
+            onlyRaw = false,
+            showTableNames = false,
+            customLabels = {},
+            columnOrder = [],
+            hiddenFields = [],
+            attachmentDownloadName,
+        } = options;
+
+        // Process fields and generate headers using shared utility
+        const { sortedFieldIds, headers } = processFieldsForExport(fields, {
+            showTableNames,
+            customLabels,
+            columnOrder,
+            hiddenFields,
         });
+
+        // Create temporary file
+        const tempFilePath = ExcelService.createTempFilePath('direct');
+
+        try {
+            // Step 1: Get source stream
+            const resultsStream = await storageClient.getDowloadStream(
+                resultsFileName,
+            );
+
+            // Step 2: Stream JSONL data to Excel temp file
+            const rowCount = await ExcelService.streamJsonlToExcelFile(
+                resultsStream,
+                tempFilePath,
+                headers,
+                fields,
+                onlyRaw,
+                sortedFieldIds,
+            );
+
+            Logger.debug(`Processed ${rowCount} rows for Excel export`);
+
+            // Step 3: Stream temp file directly to S3 (no memory spike!)
+            const fileUrl = await storageClient.uploadFileStream(
+                formattedFileName,
+                tempFilePath,
+                {
+                    contentType:
+                        'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
+                    attachmentDownloadName: attachmentDownloadName
+                        ? `${attachmentDownloadName}.xlsx`
+                        : undefined,
+                },
+            );
+
+            // Clean up temp file after successful upload
+            ExcelService.cleanupTempFile(tempFilePath);
+
+            return {
+                fileUrl,
+                truncated: false,
```

Comment:
Didn't we forget to limit lines here and return true if limit hit?
`maxLines: 1_000_000, // Excel limit - 1 million rows`